docs/design-documents/promtail-scraping-objectstore.md

```diff
@@ -0,0 +1,89 @@
+# Promtail-Objectstore-scraping
+
+Promtail currently supports scraping logs from local filesystem logs, syslog and systemd/journal. This covers most of the usecases where an user would potentially store logs. In addition to this there are few other usecases where the logs can be stored in an Object store and user would like to collect and aggregate those logs in Loki. AWS loadbalancer is an one such example, Loadbalancer access logs are stored in S3. It would be good addition to Promtail to support scraping logs from Object store to cover similar user cases mentioned above. 
+
+Cortex already has the [implementation](https://github.com/cortexproject/cortex/blob/master/pkg/chunk/aws/s3_storage_client.go) for Objecstore including S3 and GCS. We can try to re-use most of it. Still, we have to develop code for `file watching` and updating the `positions.yaml` file.  Following are the few challenges which we should try to address.
+
+### There is no appending of lines to file in object store
+
+In an object store there is no appending of lines to a file in bucket. Any file modified will be newly 
+uploaded to the bucket with the same name. So, we would have to read the entire file again from the begining since we will not be aware if only new lines were appended to the modified file or the file is completely having new content
+
+### Where should we keep positions.yaml file ?
+
+Since, the files we scrape is remotely placed in a objecstore, should we keep the `positions.yaml` also in the same place? Advantage of keeping the `positions.yaml` within the objecstore will help us to recover from the state where the scraping was left when promtail is running in docker container and the container crashes. However, `user` reading the files from the Object store would also need write access to the Objecstore. Getting write access may not be possible in some cases where people doesn't want to risk modifying the files by giving write access.
```

Comment:
You gotta be very careful with this due to the data consistency model in object storage systems. You may get stale data on a read after write, meaning you scrape the same data twice. Potentially consider a highly consistent system like consul/etcd, similar to how Loki uses these for its distributor/ingester ring. 

This would also enable usage of locking/leasing so that multiple promtail instances could coordinate scraping. I think this is pretty important since it's possible during rolling updates that you end up with multiple promtail instances running, and this could also lead to duplicated scrapes/data.

Reply:
Very valid point. Especially with S3 We should be very careful for any operation we do like list, put, get because of eventual consistency issues.

Reply:
I think this is fine for now if we just skip through out of order and try to always snapshot the position file when terminating and from time to time. This file should be just used for resuming/starting.

Reply:
True. You still need to think about concurrency though.